sql/binlog.h

```diff
@@ -879,10 +880,32 @@ class MYSQL_BIN_LOG : public TC_LOG {
     It is called by the threads (e.g. dump thread, applier thread) which want
     to read hot log without LOCK_log protection.
   */
-  my_off_t get_binlog_end_pos() const {
+  my_off_t get_binlog_end_pos(bool need_lock = false,
+                              bool wait_for_ack = false) {
     mysql_mutex_assert_not_owner(&LOCK_log);
+    if (wait_for_ack) {
+      if (need_lock) lock_binlog_end_pos();
+      const char *log_name = log_file_name + dirname_length(log_file_name);
+      const int res = strcmp(log_name, last_acked.first.c_str());
+      const my_off_t last_acked_pos = last_acked.second;
+      if (need_lock) unlock_binlog_end_pos();
+      if (res == 0) return last_acked_pos;
+      if (res > 0) return 0;
+    }
     return atomic_binlog_end_pos;
   }
+
+  void signal_semi_sync_ack(const char *const log_file,
+                            const my_off_t log_pos) {
+    const auto acked = std::make_pair(std::string(log_file), log_pos);
+    lock_binlog_end_pos();
+    if (acked > last_acked) {
+      last_acked = acked;
+      signal_update();
```

Comment:
We could update only if rpl_wait_for_semi_sync_ack is enabled. Maybe that would be slightly more performant?